Re-authentication for stronger validation to change email address.

chdinesh1089 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
https://chat.zulip.org/#narrow/stream/3-backend/topic/Re-authentication.20for.20changing.20email

**Testing Plan:** <!-- How have you tested? -->


**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

chdinesh1089 committed:
auth: Redesign email change flow.

Previously, email change was very simple but was not highly secured
as the user just needs to enter a new email in user settings and
just click the link sent to the new email to change email.

To make it more secure, we decided to re-authenticate the user.
In other words, when the user clicks the confirmation link sent
to the new email, we render login page and require the user to
login with the old email for one last time with the old email
before actually changing the email.

This commit adds support for re-authentication through
email+password auth and social auth.

To achieve that, This commit does the following:

* Adds a new django view `start_email_change` which is used to
redirect to login page with the confirmation key as a get param
named as action_key and with preview=true so that user doesn't
get redirected to home if already logged in on clicking the
confirmation link.

* The confirmation key is plumbed into the login page template
as `action_key` which is acheived by sending `action_key` as
`extra_content` to `login_page` view.

* `is_preview` flag in `login_page` is set to True again even if
action key is present in POST the user should still be able to
continue the process without being taken to home if logged in
even after making a mistake with the email+password form.

* Adds a mixin which contains all the tests and depends on
a function called do_email_change which is implemented in
the test class to run all the tests in different auth flows.
This reduces the amount of code duplication to a great extent.

* The mixin is implemented in SocialAuthBase, TestZulipRemoteUserBackend
of test_auth_backends.py and a new test class EmailChangePasswordAuthTest
in test_email_change.py.

* Removes some of the existing tests that won't work after this change.
Along with those, it also removes the end to end test as we already
test the first part of the flow in EmailChangeTestCase and other
part is being tested extensively for all supported authentication
methods.

chdinesh1089 commented on zerver/views/auth.py:
```diff
@@ -106,6 +115,39 @@ def create_preregistration_user(email: str, request: HttpRequest, realm_creation
         full_name_validated=full_name_validated,
     )
 
+def confirm_email_change(request: HttpRequest, confirmation_key: str,
```
I moved this here because of some import errors caused due to importing `auth.py` functions in `user_settings.py` and vice versa.

mateuszmandera commented on zerver/views/user_settings.py:
```diff
@@ -74,6 +74,18 @@ def confirm_email_change(request: HttpRequest, confirmation_key: str) -> HttpRes
+    except ConfirmationKeyException as exception:
+        return render_confirmation_key_error(request, exception)
+
+    logout(request)  # In case user is logged in, this avoids the redirect home.
```
I don't think we want to log out the user just for clicking the link, that might be pretty annoying to them. We should figure out how to run the re-auth flow even if they already have a logged in session.

chdinesh1089 replied:
Didn't notice earlier that we have an `is_preview` that we could use to not redirect the user to home. Added a commit for that particular change.

(We'll squash the commits at the end.)

chdinesh1089 commented on templates/zerver/login.html:
```diff
@@ -10,7 +10,12 @@
     <div class="inline-block">
         <div class="lead">
-            <h1 class="get-started">{{ _("Log in to Zulip") }}</h1>
+            <h1 class="get-started">
+                {{ _("Log in") }}
+                {% if email_change %} {{ _("to confirm email change") }} {% else %}
+                {{ _("to Zulip") }}
+                {% endif %}
```
This is giving the raw html content to be like this containing spaces and `\n`'s
```
Log in\n                \n                to Zulip\n  
```
which inturn is causing a test(`login_login_preview` in `test_auth_backends.py`) fail that asserts that `Login in to Zulip` is present.

(It looks normal in the rendered HTML. Here's a screenshot to help us decide we actually want it there
![image](https://user-images.githubusercontent.com/44665669/87704704-69237f80-c7ba-11ea-8ff9-3663de4837ef.png)

chdinesh1089 replied:
I'm confused if we should assert that `Log in\n                \n                to Zulip\n  ` is present or check if `Log in` and `to Zulip` are present or think of a different place to have that text in?

chdinesh1089 replied:
Changed it to have two complete sentences rather than breaking the sentence just to save two words.

chdinesh1089 committed:
auth: Support email change with login through remote sso auth.

This being very similar to socail auth flow, most of the
required code has already been added by previous commit.
So, this commit just sends `action_key` as param
to the functions related to this auth flow.

chdinesh1089 commented on zerver/forms.py:
```diff
@@ -379,6 +383,15 @@ def clean(self) -> Dict[str, Any]:
+                try:
+                    email_change_object = get_object_from_key(action_key, Confirmation.EMAIL_CHANGE)
+                except ConfirmationKeyException:
+                    raise ValidationError("Invalid key")
```
In a normal flow, this should not occur as we redirect the user to appropriate error page on clicking confirmation link with this code https://github.com/zulip/zulip/pull/15805/commits/8b5450c534d443474a4377f306d883e807bff0fa#diff-2a3f35796feeb6b0ded8f5b2eb2a58f0R78-R81

So, I thought just saying "Invalid key" would be enough.

chdinesh1089 commented on zerver/views/auth.py:
```diff
@@ -757,6 +757,10 @@ def login_page(
         # We want to authenticate user while letting to be logged in.
         is_preview = True
 
+    if 'action_key' in request.POST:
+        is_preview = True
+        extra_context['action_key'] = request.POST['action_key']
```
This code is for a special case i.e If the user is logged in, and clicks the link to re-authenticate then if for some reason the user gets a validation error we generally redirect to \login with some form error. The redirect takes user to home page as the user is already loggeed in. So, this code takes care of restricting that redirect to home page

mateuszmandera commented:
I didn't get bogged down in implementation details yet, because I'm guessing a bunch of that is going to change in the course of cleanups and testing - but the general direction seems right to me.

Some notes on commit structure:
1. The commit moving the function to another file due to circular imports should be a preparatory commit, so that your first commit isn't actually broken (due to the circular import issue).
2. The first commit should probably be squashed with the email+password login one (or the social flow one) because right now it's basically implementing half the process and leaves it a non-usable state, which was a bit confusing to read. While having the entire process implemented for at least one auth method in a single commit will give a much better view of what's going on when  reading the commit.

I think the goal right now should be to get this tested and cleaned up (there are still some awkward redundant import line changes and other minor style things that I didn't dive into, because a bunch of it you'll probably change anyway when looking through this again and polishing) so that we can have Tim give this an initial review. I think you can leave Remote User SSO for now, maybe move it to be the last commit, as that's not an important auth method to focus on right now and we can test it once we know we have the system right for the important methods.

mateuszmandera commented on zerver/views/auth.py:
```diff
@@ -123,6 +127,7 @@ def confirm_email_change(request: HttpRequest, confirmation_key: str) -> HttpRes
         'new_email': new_email,
         'old_email': old_email,
     }
+    do_login(request, user_profile)
```
This is to improve the user experience, so they don't have auth again right after they just authed, right? I think that makes sense, so we might want to give it more extra thought about how safe this will be, inside a function like this. Can you move this change to its own commit, as the last commit?

chdinesh1089 replied:
Yes, (though it is unnecessary if the user is already logged in.) I do not see any security issue as we call this function at two places where user is confirmed to have been authed.
* We can be sure with social auth and remote sso flows because this is being called in `login_or_register_remote_user` which is only called after successful auth afaik.
* I'm not 100% sure about the email auth case as you pointed out that not having `SimpleTemplateResponse` isn't a better check to assert that the user is authenticated which I'm convinced that it could be unsafe.

If we're convinced to change a user's email(which I think, is a stronger action than logging in) we can log them in at the same time i.e just after changing email.
Made it a separate commit.

chdinesh1089 replied:
For email+password auth case, now I've changed it to check `request.user.is_authenticated` before proceeding to `confirm_email_change` so I think we are on a safe side to log user in or take to `confirm_email_change`.

mateuszmandera commented on zerver/views/auth.py:
```diff
@@ -96,15 +97,18 @@ def get_safe_redirect_to(url: str, redirect_host: str) -> str:
 
-def confirm_email_change(request: HttpRequest, confirmation_key: str) -> HttpResponse:
+def confirm_email_change(request: HttpRequest, confirmation_key: str,
+                         user_profile: UserProfile) -> HttpResponse:
```
should we name this `acting_user` instead?

chdinesh1089 replied:
Not sure. But better than `user_profile` as it could be confusing sometimes. So changed it.

mateuszmandera commented on zerver/views/auth.py:
```diff
@@ -312,6 +326,16 @@ def login_or_register_remote_user(request: HttpRequest, result: ExternalAuthResu
+    try:
+        confirmation_object = get_confirmation_object(key)
+    except Confirmation.DoesNotExist:
+        return render(request, 'confirmation/link_does_not_exist.html')
```
this block is basically duplicating what happens inside confirm_email_change I think? We should clean this up somehow.

chdinesh1089 replied:
Not exactly. In `confirm_email_change`, we knew what `type` the confirmation object was so we were using a very similar function `get_object_from_key` which requires us to pass the object type which in `confirm_email_change`'s case was `EMAIL_CHANGE` but here we cannot know what type of object is stored in the key before we get it once we start using action key for other purposes. So, I added this function just for the sake of knowing what type the object is.

I think we should instead define a `get_confirmation_object_type`. will change it so that the code will be more readable.

mateuszmandera replied:
Can't we simply solve this weird duplicate fetching and validation of the key in both function by making `confirm_email_change` take the already fetched `Confirmation` object as argument rather than the key string?

chdinesh1089 replied:
That'd be nice but the existing `get_object_from_key` requires `confirmation_type` to be sent to get the object. If we change it to not require `confirmation_type`, we need to change it in 9 places(excluding settings) it's used. We can define another similar function but it'd be same as the existing `get_object_from_key`.

What would be a better thing to do here? I'm inclined to changing `get_object_from_key` requiring `confirmation_type` but not sure if there would be an issue doing that.

chdinesh1089 replied:
As discussed, I changed `handle_action_key` to just get the object specifying the confirmation key type to be email change as it's the only type of action key this flow supports since we can extend the `handle_action_key` later when we add other types of confirmation keys to be `action_key`.

mateuszmandera commented on zerver/views/auth.py:
```diff
@@ -312,6 +326,16 @@ def login_or_register_remote_user(request: HttpRequest, result: ExternalAuthResu
+        return render(request, 'confirmation/link_does_not_exist.html')
+
+    if confirmation_object.type == Confirmation.EMAIL_CHANGE:
+        return confirm_email_change(request, key, user_profile)
```
this isn't returning a response if the type is not `EMAIL_CHANGE`. Is mypy not catching this?

chdinesh1089 replied:
mypy isn't catching it.(Not sure why but I guess all current code paths either give an invalid key or one that is of type EMAIL_CHANGE). Will add a `else` here and render `'confirmation/link_does_not_exist.html'`. Would it be fine or shall I make a new page for this case explaining what's the issue? (That should generally not happen I think but we should handle that case too)

mateuszmandera commented on zerver/views/auth.py:
```diff
@@ -719,6 +750,22 @@ def login_page(
+    extra_context["next"] = next
+
+    if extra_context.get("action_key") is None:
+        extra_context["action_key"] = ''
```
do we need this? Can't action_key simply be absent in the dictionary?

chdinesh1089 replied:
I initially tried it without this but the template render is skipping it without rendering i.e without this the HTML looks like `action_key = "{{ action_key}}"`. I think Jinja2 should render it as an empty string but I'm not sure why this is being skipped. So, as a kind of workaround, I added this code.

mateuszmandera replied:
Will be cleaner to add a conditional in the template to just skip this thing from the form if not action_key.

mateuszmandera commented on zerver/views/auth.py:
```diff
@@ -719,6 +750,22 @@ def login_page(
+        is_preview = True
+
+    action_key_from_post = request.POST.get('action_key')
+    action_key_exists_in_post  = action_key_from_post and action_key_from_post is not None
+    if action_key_exists_in_post:
```
This seems redundant, all we need is
```suggestion
    if action_key_from_post:
```
since it does exactly thhe same

chdinesh1089 replied:
Ohh, yes, it's redundant. I was a bit confused here when puppeteer test failed and ended up doing this. changed it.

mateuszmandera commented on zerver/views/auth.py:
```diff
@@ -772,7 +817,10 @@ def login_page(
         # https://github.com/django/django/blob/master/django/template/response.py#L19.
         update_login_page_context(request, template_response.context_data)
-
+    else:
```
A simple `else` here seems extremely unsafe. You're assuming anything else than `SimpleTemplateResponse` means the user authenticated successfully. Even if that's true right now, it might not be so for example if Django just change the base class to something else. Or if there's possibility of something else being returned in certain cases of the user failing to auth. We need to find out what exactly the expected `template_response` is in case of success and verify it here.

chdinesh1089 replied:
It is just a `HttpResponseRedirect`(added a print, also verified [here](https://github.com/django/django/blob/3f2821af6bc48fa8e7970c1ce27bc54c3172545e/django/contrib/auth/views.py#L93)) not sure if it'd be safer but it's better than the current code. So added that check too.

chdinesh1089 replied:
A stronger check here would be `request.user.is_authenticated`. So changed it check for that and continue to `confirm_email_change`.

mateuszmandera replied:
Great, `request.user.is_authenticated` seems like exactly the kind of thing we want. Mere fact of returning `HttpResponseRedirect` wouldn't inspire me with confidence that the user successfully authenticated.

chdinesh1089 committed:
auth: Log in user on successful email change.

We authenticate the user before changing their email address.
To improve user experience by not having them auth again right after
they just authed, we simply log them in just after changing the
email address.

Also adds a link to redirect user to homepage in
`confirm_email_change.html`.

mateuszmandera commented on templates/confirmation/confirm_email_change.html:
```diff
@@ -11,6 +11,7 @@ <h1 class="get-started">{{ _('Email changed!') }}</h1>
         <div class="app-main white-box">
             This confirms that the email address for your Zulip account has changed
             from <a href="mailto:{{old_email}}">{{old_email}}</a> to <a href="mailto:{{new_email}}">{{new_email}}</a>.
+            <p><a href="\">Click here</a> to go to homepage.</p>
```
```suggestion
            <p><a href="/">Click here</a> to go to homepage.</p>
```

mateuszmandera commented on zerver/views/auth.py:
```diff
@@ -290,6 +290,14 @@ def login_or_register_remote_user(request: HttpRequest, result: ExternalAuthResu
+
+    action_key = result.data_dict.get('action_key')
+    if action_key:
+        if user_profile is None:
```
if you put this block a few lines below, it would get much simpler, as whether user needs to register is checked and handled right below (and it handles the mirror dummy case too, which also requires registering and was missed here.

chdinesh1089 replied:
Do we want to raise a JSON error or just take the user to registration page? Raising an error seems right to me as it could avoid some confusion if the user accidentally logs in with some other account. So, for now I made it raise json error. we can change it if you think it's not ideal.

mateuszmandera replied:
Hmm okay I think you're right, the user might get confused and log in with the new email rather than old, taking them to registration would be weird and an error is better. Json error is not ideal, it's be better experience to take the user to the login page again with the action key and an appropriate message - but doing that is more work so let's keep just the json error for now and make error handling nicer in followups. 

But `user_profile.is_mirror` needs to be accounted for, and probably you can make this login join with that block below that has the same `if` in a way that'll read nicer.

chdinesh1089 replied:
Yes, I added an `if` inside `if user_profile is None or user_profile.is_mirror_dummy:` to check for action key and raise error.  Now looking at it again, I felt like keeping `action_key` related code as one block in the code would look clean. So thought I'd modify like this, but my doesn't let me(see [this post](https://chat.zulip.org/#narrow/stream/92-learning/topic/mypy.20error/near/968696) in chat.zulip.org)
```python
....
    user_needs_to_register = user_profile is None or user_profile.is_mirror_dummy

    action_key = result.data_dict.get('action_key')
    if action_key:
        if user_needs_to_register:
            raise JsonableError(_("Incorrect login. Login with your old email to change your email address"))
        return handle_action_key(request, user_profile, action_key)

    if user_needs_to_register:
        return register_remote_user(request, result)
    # Otherwise, the user has successfully authenticated to an
...
```

mateuszmandera commented on zerver/views/auth.py:
```diff
@@ -719,6 +743,20 @@ def login_page(
     # that disables the default behavior of redirecting logged-in users to the
     # logged-in app.
     is_preview = 'preview' in request.GET
+
+    extra_context = kwargs.pop('extra_context', {})
+    extra_context["next"] = next
+
+    action_key_from_post = request.POST.get('action_key')
+    if action_key_from_post:
+        # In case some error occurs in the form validation,
+        # user should still be able to retry.
+        extra_context['action_key'] = action_key_from_post
+
+    if extra_context.get("action_key") is not None:
+        # We want to authenticate user while letting the user to be logged in.
+        is_preview = True
+
```
```suggestion
    extra_context = kwargs.pop('extra_context', {})
    extra_context["next"] = next

    action_key = request.POST.get('action_key')
    extra_context['action_key'] = action_key
    if action_key:
        # We want to authenticate user while letting the user to be logged in.
        is_preview = True
```

reads simpler and cleaner. Detailing that the key comes from POST in the variable name seems rather unnecessary too.

chdinesh1089 replied:
The key only comes from POST only during email+password once the user clicks login button otherwise we send action key in through `extra_context` in the `start_email_change` view that's run on clicking confirmation link.

Doing that would replace the `action_key` with `None`.

Can we do something in the code to make this clear? 🤔

mateuszmandera replied:
Ah I see. Hmm well either way, the natural role `extra_context`  is to simply affect how the template will be rended. What you're doing with 
```
    if extra_context.get("action_key") is not None:
        # We want to authenticate user while letting the user to be logged in.
        is_preview = True
```

is altering the behavior of the login_page function with what's supposed to only affect the template. That's not a great way of conveying "don't redirect the user to home" to the function. One option is to not call `login_page` with extra_context in start_email_change, but rather redirect the user to `/login/` with `action_key` param in the `GET` and grab it from there.

chdinesh1089 replied:
I like the idea of redirecting the user to `/login/` with `action_key` as a `GET` param and grabbing it from there. This also resolves the circular import error we had as we now do not want to import `login_page` from `auth.py`. So, I've removed that commit.

mateuszmandera commented on zerver/views/auth.py:
```diff
@@ -96,28 +97,26 @@ def get_safe_redirect_to(url: str, redirect_host: str) -> str:
-    if user_profile.realm.email_changes_disabled and not user_profile.is_realm_admin:
+    if acting_user != email_change_object.user_profile:
+        raise JsonableError(_("Incorrect login. Login with your old email to change your email address"))
+
```
let's just set `user_profile = email_change_object.user_profile` here after the validation annd keep using that vairable below.

mateuszmandera commented on zerver/lib/test_helpers.py:
```diff
@@ -598,3 +602,27 @@ def zulip_reaction_info() -> Dict[str, str]:
+def create_email_change_confirmation_key(user_profile: UserProfile, new_email: str,
+                                         email_address_visibility: int=Realm.EMAIL_ADDRESS_VISIBILITY_EVERYONE,
+                                         date_sent: Optional[datetime]=None) -> str:
+    do_set_realm_property(
```
A function named like this shouldn't sneakily change realm properties. If tests need email visibility changed (why btw?) they should do it explicitly. Also, seems like `create_email_change_confirmation_key` is the kind of code that should be shared between tests and the production code that does the same thing, rather than having two places in the code that aim to do the same thing.

chdinesh1089 replied:
> why btw?

There's a test in the previous tests called `test_change_delivery_email_end_to_end_with_admins_visibility`. It is for that and the very few similar tests. Changed it to have it in the actual test's code.

I think we cannot really use the production code. We create the URL and mail it to the user in a single function  `do_start_email_change_process`. So we cannot directly get the URL/key from production code. and another issue would be us setting the required date to test expired key.

chdinesh1089 replied:
Ahh, we default to `EMAIL_ADDRESS_VISIBILITY_ADMINS`, now removing that is causing the coverage issue 😬

chdinesh1089 replied:
Changed `test_social_auth_email_change_with_admins_visibility` to `test_social_auth_email_change_delivery_email_public` and moved the assertion that checks if `email` isn't changed and remained as user id to the `test_social_auth_email_change_success` test.
[Here](https://pastebin.com/kUQrz9Uz)'s the diff for that fix.

mateuszmandera replied:
> I think we cannot really use the production code. We create the URL and mail it to the user in a single function do_start_email_change_process. So we cannot directly get the URL/key from production code. and another issue would be us setting the required date to test expired key.

I'd say we should extract the common "create the link" logic into a function, so that it can used in both `start_email_change_process` and in tests, it can take `date_sent` as an argument.

chdinesh1089 replied:
"create the link" is already a different function(`create_confirmation_link`) which generates the key and returns URL with key in it. 
For quick reference here's the production code
```python
def do_start_email_change_process(user_profile: UserProfile, new_email: str) -> None:
    old_email = user_profile.delivery_email
    obj = EmailChangeStatus.objects.create(new_email=new_email, old_email=old_email,
                                           user_profile=user_profile, realm=user_profile.realm)

    activation_url = create_confirmation_link(obj, Confirmation.EMAIL_CHANGE)
    from zerver.context_processors import common_context
    context = common_context(user_profile)
    context.update({
        'old_email': old_email,
        'new_email': new_email,
        'activate_url': activation_url,
    })
    language = user_profile.default_language
    send_email('zerver/emails/confirm_new_email', to_emails=[new_email],
               from_name=FromAddress.security_email_from_name(language=language),
               from_address=FromAddress.tokenized_no_reply_address(),
               language=language, context=context,
               realm=user_profile.realm)
```
I'm unable to see a way to break that so we could use it for test code which would let us set our own `date_sent`. :disappointed:

mateuszmandera replied:
Call `create_confirmation_link` in the test helper, and change `date_sent` field on the created object to the intended one? :thinking:

chdinesh1089 replied:
Hmm, that way we'd just save this statement 
```
    Confirmation.objects.create(content_object=obj,
                                date_sent=date_sent,
                                confirmation_key=key,
                                type=Confirmation.EMAIL_CHANGE)
```
in(just for reference)
```python
def create_email_change_confirmation_key(user_profile: UserProfile, new_email: str,
                                         date_sent: Optional[datetime.datetime]=None) -> str:
    old_email = user_profile.delivery_email
    new_email = new_email
    obj = EmailChangeStatus.objects.create(new_email=new_email,
                                           old_email=old_email,
                                           user_profile=user_profile,
                                           realm=user_profile.realm)
    key = generate_key()
    if date_sent is None:
        date_sent = timezone_now()
    Confirmation.objects.create(content_object=obj,
                                date_sent=date_sent,
                                confirmation_key=key,
                                type=Confirmation.EMAIL_CHANGE)
    return key
```
and will have to add a line to retrieve `action_key` from the link to send it in as GET/POST's for auth flows along with adding the code for changing `date_sent`.

mateuszmandera replied:
```
    obj = EmailChangeStatus.objects.create(new_email=new_email, old_email=old_email,
                                           user_profile=user_profile, realm=user_profile.realm)

    activation_url = create_confirmation_link(obj, Confirmation.EMAIL_CHANGE)
```

this can be extracted from the prduction code into a `create_email_change_confirmation_link`  helper, and then the test will call that create_email_change_confirmation_link helper and alter the date_sent and grab the key :thinking:  It might not save many, or any, actual lines of code but the point is to encapsulate the details of the procedure of creating this object, so I think it's worth it nonetheless

chdinesh1089 replied:
> It might not save many, or any, actual lines of code but the point is to encapsulate the details of the procedure of creating this object, so I think it's worth it nonetheless


Now, it very clear to me why we wanted to use the production code. Thanks, made the changes as a prep commit.

mateuszmandera commented on zerver/tests/test_email_change.py:
```diff
@@ -1,22 +1,31 @@
 
+def get_email_change_confirmation_url(user_profile: UserProfile, new_email: str, key: str,
+                                      email_address_visibility: int=Realm.EMAIL_ADDRESS_VISIBILITY_EVERYONE,
+                                      ) -> str:
```
With the previous test helper, this is now adding the 3rd place in the code with the same aim of creating email change confirmation url, we don't want that. `email_address_visibility` seems like a redundant arg here?

chdinesh1089 replied:
Agreed. And all the checks which were that the login page was loaded were also redundant as they are already tested in `EmailChangeTestCase`. So, I've imported `create_email_change_confirmation_key` and just used the key to send POST requests.

Yes, `email_address_visibility ` is redundant. Seems like I forgot to add code specific to that. I'm surprised lint didn't complain. (removed that function now.)

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1471,6 +1476,305 @@ def test_social_auth_invited_as_admin_but_expired(self) -> None:
+        result = self.client_get(result.url)
+        self.assert_in_success_response(["This confirms that the email address for your Zulip"],
+                                        result)
+        user_profile.refresh_from_db()
+        user_profile = get_user_by_delivery_email(new_email, user_profile.realm)
+        self.assertTrue(bool(user_profile))
+        user_profile = get_user_profile_by_id(user_profile.id)
+        self.assertEqual(user_profile.email, f"user{user_profile.id}@zulip.testserver")
+        with self.assertRaises(UserProfile.DoesNotExist):
+            get_user_by_delivery_email(old_email, user_profile.realm)
+
+        email_change_object = get_object_from_key(action_key, Confirmation.EMAIL_CHANGE)
+        email_change_object.refresh_from_db()
+        self.assertEqual(email_change_object.status, 1)
```
This ratjer repetetive logic across the tests should probably be in some kind of `verify_user_email_changed` helper. I see that this is how the old email change tests were already done, but I don't understand why this validation is done in such a roundabout way - wouldn't it be sufficient to check `user_profile.delivery_email` is the correct new value, and that `status` of the email change object is changed and that's all? :thinking:

chdinesh1089 replied:
Hmm, I guess checking if there's no account exists with the old email is worthy. The assertion to the non-delivery email is only done in `test_social_auth_email_change_delivery_email_with_admins_visibility` which I think is the whole point of the test. 

`verify_user_email_changed` sounds like a good idea. But I'm confused where to put that function in as it needs to be in a test subclassing `TestCase` for us to be able to use `self.assertEqual` and similar functions and we have 3 places(3 different test classes) we want to use it. I don't think it's a good idea to put it in `ZulipTestCase`. 

What do you think would be a good solution for this so that we can avoid duplication?

mateuszmandera replied:
Hmm, why would it be bad to put it in zuliptestcase?

chdinesh1089 replied:
We would use `verify_user_email_changed` for just a very few tests cases and `ZulipTestCase` is being subclassed for every test class in our test code and most of them do not need it. It didn't feel like a good idea to me to put a function that has no use in most of the subclasses it stays hidden.
- - -
Now I re-wrote all these tests into mixin class and only the part that differs in each flow is extracted into a different function(`do_email_change`) which needs to be defined in the classes implementing this mixin. This reduced the amount of test code to one-third of what we had earlier.

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1471,6 +1476,305 @@ def test_social_auth_invited_as_admin_but_expired(self) -> None:
+
+        with self.assertRaises(UserProfile.DoesNotExist):
+            get_user_by_delivery_email(new_email, user_profile.realm)
+        # No exception should raise.
```
```suggestion
        # No exception should be raised.
```

Though I'm not sure if that kind of comment isn't rather redundant :thinking: We can clearly see from the code that exceptions aren't expected to happen here :grinning:

chdinesh1089 replied:
Thought no one should wonder why's there `get_user_by_delivery_email(old_email, user_profile.realm)` that isn't being used in any way. But yes, it seems to be a bit obvious seeing the two lines just above this and in the context of tests it is expected as far as I can see. So, removed the comments 😀

mateuszmandera commented:
> Changes the end to end test to only check till the login page
template is loaded as new tests are added that test successful
email change from confirmation link in the coming commits. So,
renamed the test appropriately.

This in the commit message is no longer accurate i think

chdinesh1089 commented:
(Made a few refinements to the tests(like comments etc.) after a self-review)

mateuszmandera commented on zerver/tests/email_change_test_helper.py:
```diff
@@ -0,0 +1,206 @@
+                                type=Confirmation.EMAIL_CHANGE)
+    return key
+
+class EmailChangeTestMixin(ZulipTestCase):
```
ohh, this is really cool, kinda like the social auth flow test base class. I didn't expect this would generalize into a base class so nicely. Though  I think this should go to `test_classes.py`, we have some other general "library" classes there.

chdinesh1089 replied:
sure, glad you liked it, moving it now. Does the name sound good? or do we want to go for something like `EmailChangeTestBase` or `EmailChangeBase`?

chdinesh1089 replied:
Moved it to `test_classes.py`.

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1472,6 +1479,30 @@ def test_social_auth_invited_as_admin_but_expired(self) -> None:
 
+    def do_email_change(self, user_profile: UserProfile, action_key: str,
+                        different_account: bool=False) -> HttpResponse:
+        if different_account:
+            user_profile = self.example_user('cordelia')
```
```suggestion
        if different_account:
            assert user_profile != self.example_user('cordelia')
            user_profile = self.example_user('cordelia')
```

to avoid weird bugs if somebody passes cordelia as the user_profile argument

chdinesh1089 replied:
Didn't think about that, interesting, Thanks! Added the suggestion.

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -2820,6 +2857,30 @@ def test_github_unverified_email_with_existing_account(self) -> None:
+    def test_social_auth_email_change_unregistered_user_login(self) -> None:
+        # Overriden as user is taken to choose email screen and
+        # it needs to be handled by send `expect_choose_email_screen`
+        # to `social_auth_test`.
```
is it not possible to write the `do_email_change` function in such a way that this override won't be needed?

chdinesh1089 replied:
We'd need to have conditionals to check if is github auth and add some code to check that the user doesn't exist which means that it most probably is this particular test for unregistered user, then we need to set `expect_choose_email_screen`. I think this would clutter up `do_email_change`. One other thing we could do is send some param `unregistered_user=True` to avoid some code and add an `if backend == GitHubAuthBackend and unregistered` in the `do_email_change` and then set `expect_choose_email_screen` to True. But this param would remain useless for other auth flows and become confusing to someone reading the code if we introduce an argument for `do_change_email` for the purpose of a single test.

Both of those options seemed not optimal to me and had to write this overridden function.

mateuszmandera replied:
> then we need to set expect_choose_email_screen. I think this would clutter up do_email_change

That sounds like simply an extra argument and a 2-liner `if` case in `do_change` though I think? Or would it be more clutter than that

chdinesh1089 replied:
> That sounds like simply an extra argument and a 2-liner `if` case in `do_change` though I think?

Yes, if we give an extra argument it's just a simple if in `do_change`. But I'm not sure if it's fine to send an arg just for Github where the test just doesn't run in social auth flow. Is it fine?

chdinesh1089 replied:
> That sounds like simply an extra argument and a 2-liner `if` case in `do_change` though I think?

Yes, if we send an argument it's just a simple `if`, but not sure if it's fine to send it as an arg from the test only for GIthub auth where it is being run in 2 other test classes too. Is it fine?

chdinesh1089 replied:
we could do this in `do_email_change` to avoid sending an arg, what is your opinion on this. (This seems like clutter to me but maybe not)
```diff
    def do_email_change(self, user_profile: UserProfile, action_key: str) -> HttpResponse:
         email = user_profile.delivery_email
         name = user_profile.full_name
 
+        expect_choose_email_screen = False
+        if self.backend == GitHubAuthBackend:
+            try:
+                get_user_by_delivery_email(email, user_profile.realm)
+            except UserProfile.DoesNotExist:
+                # The test is testing unregistered user logging in
+                # to change email. In such case, github auth takes to
+                # choose email screen.
+                expect_choose_email_screen = True
+
         account_data_dict = self.get_account_data_dict(email=email,
                                                        name=name)
 
         result = self.social_auth_test(account_data_dict, subdomain='zulip',
-                                       action_key=action_key)
+                                       action_key=action_key,
+                                       expect_choose_email_screen=True)
```

mateuszmandera commented on zerver/views/development/email_log.py:
```diff
@@ -112,10 +112,7 @@ def generate_all_emails(request: HttpRequest) -> HttpResponse:
-
-    # Reset the email value so we can run this again
-    do_change_user_delivery_email(user_profile, registered_email)
+    assert result.status_code == 302
```
what's the story behind this change?

chdinesh1089 replied:
We decided to redirect user to login page with a GET param but this is expecting a status code of 200 and with this change in email change flow, the email won't change just on sending a request to that link. So, removed the `do_change_user_delivery_email` too.
Mentioned this in commit message too
```
In zerver/views/development/email_log.py, changed the
assertion of status to 302 as we redirect to login page
with this commit on clicking email change confirmation link.
Also removed the change of delivery email as it's no longer
needed since the email won't be changed on clicking the link.
```

chdinesh1089 replied:
Changed it to authenticate using email+password.

mateuszmandera replied:
Have you tested manually whether this works? It looks right, but worth checking.

chdinesh1089 replied:
yes, tested manually before pushing. it works :)

chdinesh1089 committed:
email_change: Extract confirmation link creation to a different function.

Acts as a prep-commit for redesigning email change process to
re-authenticate as we want to be using production code for tests
inorder to encapsulate/deduplicate the details of procedure of
creating EmailChangeStatus object.

chdinesh1089 commented:
Oops, sorry for the delay. This was tracked in a different branch `new_email_wip` locally and I pushed it to `new_email_wip` instead of pushing it here and thought this PR has been updated 😬

mateuszmandera commented on zerver/lib/test_classes.py:
```diff
@@ -1103,3 +1109,189 @@ def setUp(self) -> None:
     def setUpBeforeMigration(self, apps: StateApps) -> None:
         pass  # nocoverage
+
+
```
one whitespace too much

mateuszmandera commented on zerver/lib/test_classes.py:
```diff
@@ -1103,3 +1109,189 @@ def setUp(self) -> None:
+    authenticate based on the authentication flow.
+    '''
+    __unittest_skip__ = True
+
```
this class should have a dummy `do_Change` function definition, to show the intended function signature, with a docstring quickly explaining its intent, and doing `raise NotImplementedError` because overriding it is obligatory

chdinesh1089 replied:
Added, didn't know what the right thing to do to convey that overriding is obligatory so I initially just added a comment.

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -1472,6 +1475,31 @@ def test_social_auth_invited_as_admin_but_expired(self) -> None:
         self.assertEqual(created_user.role, UserProfile.ROLE_MEMBER)
 
+    def do_email_change(self, user_profile: UserProfile, action_key: str,
+                        different_account: bool=False) -> HttpResponse:
```
Why do we need the `different_account` argument, rather than the caller passing a different user in the `user_profile` argument instead?

chdinesh1089 replied:
Ohh, right, that argument doesn't make sense. Thanks, removed.

mateuszmandera commented on zerver/lib/test_classes.py:
```diff
@@ -1103,3 +1109,189 @@ def setUp(self) -> None:
+        user_profile = self.example_user('hamlet')
+        new_email = 'hamlet-new@zulip.com'
+
+        date_sent = timezone_now() - datetime.timedelta(days=2)
```
Where in the code is it determined that an email change key is expired? I see other types of keys have  constants such as `INVITATION_LINK_VALIDITY_DAYS` but i dont see it for email change :thinking:

chdinesh1089 replied:
It has the default validity set by this in `default_settings.py` `CONFIRMATION_LINK_DEFAULT_VALIDITY_DAYS = 1`.

>Where in the code is it determined that an email change key is expired? 

All sorts of checks(in general for all confirmation key) including this are done by `get_object_from_key`. we use it to get the email change object twice in this PR, once in `handle_action_key` and `start_email_change`(we do it in this only to validate)

mateuszmandera replied:
Okay, thanks for checking! I think it'd be better to use `CONFIRMATION_LINK_DEFAULT_VALIDITY_DAYS + 1` here, as that's clearer, rather than a magical value of 2 that will make folks puzzle over the same question i asked above :grinning:

chdinesh1089 replied:
Ohh, makes sense. Thanks, will change to use `CONFIRMATION_LINK_DEFAULT_VALIDITY_DAYS + 1`

chdinesh1089 replied:
Changed.

mateuszmandera commented on zerver/lib/test_classes.py:
```diff
@@ -1103,3 +1109,189 @@ def setUp(self) -> None:
+        result = self.do_email_change(user_profile, action_key)
+        self.assert_in_success_response(["The confirmation link has expired or been deactivated."], result)
+
+    def test_social_auth_email_change_unregistered_user_login(self) -> None:
```
a `social_auth_...` test case shouldn't be in this general mixin

mateuszmandera commented on zerver/lib/test_classes.py:
```diff
@@ -1117,3 +1123,193 @@ def setUp(self) -> None:
+        registration page social and remote auth flows and email change should happen.
+        '''
+
+        # UserProfile is created here(not stored to db) as `do_change_email`
+        # expects a UserProfile
+        unregistered_user_profile = create_user_profile(get_realm('zulip'), email='new_email@zulip.com',
+                                                        password=None, active=False, bot_type=None,
+                                                        full_name='Nonexistent User', bot_owner=None,
+                                                        is_mirror_dummy=False, tos_version=None,
+                                                        timezone=None)
```
would we be able to get rid of this hackiness if `do_change_email` took an email address (and realm?) rather than a userprofile?

chdinesh1089 replied:
Ahh, yes, that sounds like a great idea. email and action key would be sufficient remote sso and email+password. We'd need name and subdomain for social auth.
I think sending those two as **kwargs would be a good idea as we don't need them in two of the auth flows. Does it make sense to you?

chdinesh1089 replied:
Okay, changed to just send email and action key. This let us not get rid of the step of getting `email` from `user_profile` but since `name` was needed in social auth flow tests, made a `try` `except` block like this
```
    def do_email_change(self, email: str, action_key: str) -> HttpResponse:
        try:
            user_profile = get_user_by_delivery_email(email, get_realm("zulip"))
            name = user_profile.full_name
        except UserProfile.DoesNotExist:
            name = "Nonexistent User"
```

chdinesh1089 commented:
Rebased :slightly_smiling_face:
